File: cpp/observer/Car.cpp
#include "Car.h"

int Car::getPosition()
{
    return position;
}

void Car::setPosition(int newPosition)
{
    position = newPosition;
    notify();
}

void Car::attach(Observer *observer)
{
    observerList.push_back(observer);
}

void Car::detach(Observer *observer)
{
    observerList.erase(remove(observerList.begin(), observerList.end(), observer), observerList.end());
}

File: java/observer/Car.java
import java.util.ArrayList;
public class Car {
    private int position;
    private ArrayList<Observer> observerList = new ArrayList<Observer>();
    public int getPosition() {
        return position;
    }
    public void setPosition(int position) {
        this.position = position;
        notifyObserver();
    }
    public void attach(Observer observer) {
        observerList.add(observer);
    }
    public void detach(Observer observer) {
        observerList.remove(observer);
    }
    private void notifyObserver() {
        for (Observer observer : observerList) {
            observer.update();
        }
    }
}